### DISCUSSION TITLE
产品需求文档-任意进度计算器

### DISCUSSION THREAD

**Discussion by yongkangl** [HVML/PurC#74] [Ideas]
# 任意精度计算器产品需求文档

本产品“任意精度计算器”的设计目标：
- 体现HVML编程语言的特点。
- 使得任意精度计算器具有一定的实用性。

为了体现HVML本身的特点，本产品的设计宗旨如下：
- 不对界面等作特别的要求，以免影响代码阅读。

## 设计需求

为了达到基本的实用性，本计算器需必须要实现的功能：

- 输入任意精度数字并予以显示的功能。
- 对两个任意精度的数字进行基础的加减乘除运算并显示结果的功能。
- 对以往的输入和计算结果历史予以保存和调入的功能。

本产品的界面应包含以下几个功能区块：

- 输入区，主要为按钮控件，供用户输入数据。
- 显示区，主要显示用户输入和最终结果。
- 状态区，用以显示状态和上文提到要求的历史记录功能。

按照惯例，最终结果显示区应在界面的最上方，下方紧接着输入区。状态区可以处于最下方，亦可放置在显示区与输入区之间。

本产品的名称可以显示在程序的标题栏等地方，要求能够明显的被看到，但是总体显示尺寸不得大于某个功能区块的大小，以防喧宾夺主。

其他空白部位可以放置一些背景或者文字，要求实现此种放置的代码不能够太复杂以免影响代码阅读。

![overall](https://github.com/HVML/PurC/assets/112754121/f534e4b3-e1e2-4dbe-8a18-fa4a573f5bcf)

### 显示区

因为是任意精度计算器，需要一个进度条用来表示当前显示的位置。下方有一个显示框用来显示数字。显示字体要求较大 (18pt以上)。

![display](https://github.com/HVML/PurC/assets/112754121/9662d092-83d2-406d-b0d7-dfa03dbebdee)

当显示结果未超出显示范围时，进度条为锁定状态，无法使用鼠标或者热键进行操作。
当显示结果超出显示范围时，进度条需要激活。进度条可以使用鼠标进行拖动，或者使用进度条默认绑定的热键进行操作。通过操作进度条，能够使用户看到结果所有的片段，不得有死区。

### 输入区

输入区主要为按钮，包含数字键，运算键和功能键。

数字键和小数点的排列要求与小键盘排列基本相似；且数字键和加减乘除回车要与小键盘的按键进行热键绑定，以方便输入数字。

按下任意数字键和符号键进入输入过程，结果显示在显示区。
按下运算键，则当前运算被设置为按键代表的运算。如当前已有在记录的运算，则将当前缓存（历史）中的数据现有的运算操作进行运算并把结果显示在显示区，以替代之前的数字。同时将被取代的数字和运算写入历史。
按下回车则以当前记录的运算操作进行运算；如无运算符则设置 加法 为当前运算操作。

功能键-复制：将当前选中的数字以字符串的形式复制到剪贴板中。
功能键-黏贴：试图将粘贴板内容以数字的形式读取进来。

### 状态区
![status](https://github.com/HVML/PurC/assets/112754121/2bd4f2a2-4442-4a40-8763-409e88b2aa15)

状态区主要是数字和运算操作的历史记录。右边有一个进度条可供翻页，当历史记录较多时。

单击状态条的某个条目，会将选中的数字覆盖当前记录的数字，选中运算操作历史时无动作。当前显示区的数字被替换时，当前显示区的数字需要被记录为历史记录，处在最新的状态，可以影响进度条长短但是当前状态去显示不得翻页。

**Comment by VincentWei** [comment]
我觉得目前的电脑用计算器最大的问题是输入太麻烦，用鼠标或者手指一个个输入数字特别繁琐。所以，我的建议第一点就是取消键盘，直接按住说话，比如说“250乘250”，界面上就显示 250x250，再说“除以8”，界面上显示 `250x250/8`，最后说“给结果”，就显示结果。

还有一个点，就是键盘换成公式的输入键盘，可以直接输入和编辑数学公式。有了数学公式，可以搞很多其他的事情。不过这个事情就复杂了，可以搞另一个产品了。
